database/database.go

```diff
@@ -66,38 +68,86 @@ type (
-// integrating with the configured database provider.
+// New creates and returns an engine capable of integrating with the configured database provider.
 //
-// Currently the following database providers are supported:
+// Currently, the following database providers are supported:
 //
-// * Postgres
-// * Sqlite
-// .
-func New(s *Setup) (Interface, error) {
-	// validate the setup being provided
-	//
-	// https://pkg.go.dev/github.com/go-vela/server/database?tab=doc#Setup.Validate
-	err := s.Validate()
+// * postgres
+// * sqlite3
+func New(c *Config) (Interface, error) {
+	// validate the configuration being provided
+	err := c.Validate()
 	if err != nil {
 		return nil, err
 	}
 
-	logrus.Debug("creating database service from setup")
+	// create new database engine
+	e := &engine{
+		Config:   c,
+		Database: new(gorm.DB),
+		Logger:   logrus.NewEntry(logrus.StandardLogger()).WithField("database", c.Driver),
+	}
+
+	e.Logger.Trace("creating database engine from configuration")
 	// process the database driver being provided
-	switch s.Driver {
+	switch c.Driver {
 	case constants.DriverPostgres:
-		// handle the Postgres database driver being provided
-		//
-		// https://pkg.go.dev/github.com/go-vela/server/database?tab=doc#Setup.Postgres
-		return s.Postgres()
+		// create the new Postgres database client
+		e.Database, err = gorm.Open(postgres.Open(e.Config.Address), &gorm.Config{})
+		if err != nil {
+			return nil, err
+		}
 	case constants.DriverSqlite:
-		// handle the Sqlite database driver being provided
-		//
-		// https://pkg.go.dev/github.com/go-vela/server/database?tab=doc#Setup.Sqlite
-		return s.Sqlite()
+		// create the new Sqlite database client
+		e.Database, err = gorm.Open(sqlite.Open(e.Config.Address), &gorm.Config{})
+		if err != nil {
+			return nil, err
+		}
 	default:
 		// handle an invalid database driver being provided
-		return nil, fmt.Errorf("invalid database driver provided: %s", s.Driver)
+		return nil, fmt.Errorf("invalid database driver provided: %s", c.Driver)
 	}
+
+	// capture database/sql database from gorm.io/gorm database
+	db, err := e.Database.DB()
+	if err != nil {
+		return nil, err
+	}
+
+	// set the maximum amount of time a connection may be reused
+	db.SetConnMaxLifetime(e.Config.ConnectionLife)
+	// set the maximum number of connections in the idle connection pool
+	db.SetMaxIdleConns(e.Config.ConnectionIdle)
+	// set the maximum number of open connections to the database
+	db.SetMaxOpenConns(e.Config.ConnectionOpen)
+
+	// verify connection to the database
+	err = e.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	// create database agnostic engines for resources
+	err = e.NewResources()
+	if err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+// NewTest creates and returns an engine that integrates with an in-memory database provider.
+//
+// This function is ONLY intended to be used for testing purposes.
+func NewTest() (Interface, error) {
```

Comment:
This will be needed for tests all over the place that have a dependency on `database` logic

i.e. `middleware`, `secret` etc.